src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -233,6 +233,13 @@ protected virtual string GenerateUrl(string protocol, string host, VirtualPathDa
             // VirtualPathData.VirtualPath returns string.Empty instead of null.
             Debug.Assert(pathData.VirtualPath != null);
 
+            var url = FastGetUrl(protocol, host, pathData, fragment);
```

Comment:
Name this `FastGenerateUrl` or similar to make it clear that this handles a special case of `GenerateUrl`